// pogger/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = exports.debug = exports.event = exports.error = exports.warning = exports.success = exports.info = exports.log = exports.mock = void 0;
const chalk_1 = require("chalk");
const utils = __importStar(require("./utils"));
exports.utils = utils;
exports.mock = __importStar(require("./mock"));
const { infoIcon, errorIcon, successIcon, warningIcon, eventIcon, debugIcon, logIcon, } = utils.icons;
exports.log = utils.createLogger(logIcon, chalk_1.grey);
exports.info = utils.createLogger(infoIcon, chalk_1.cyan);
exports.success = utils.createLogger(successIcon, chalk_1.green);
exports.warning = utils.createLogger(warningIcon, chalk_1.yellowBright);
exports.error = utils.createLogger(errorIcon, chalk_1.red);
exports.event = utils.createLogger(eventIcon, chalk_1.magenta);
exports.debug = utils.createLogger(debugIcon, chalk_1.yellowBright);
