// pogger/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMockLogger = exports.createLogger = exports.generateDate = exports.icons = void 0;
const chalk_1 = require("chalk");
const supports_color_1 = require("supports-color");
const log = (logs, icon, color) => console.log(generateDate(), color(icon), ...logs);
exports.icons = {
    infoIcon: supports_color_1.stdout ? "ℹ" : "i",
    successIcon: supports_color_1.stdout ? "✔" : "√",
    warningIcon: supports_color_1.stdout ? "⚠" : "‼",
    errorIcon: supports_color_1.stdout ? "✖" : "×",
    eventIcon: "☄",
    debugIcon: "☢",
    logIcon: "✎",
};
function generateDate() {
    const options = {
        hour: "numeric",
        minute: "numeric",
        second: "numeric",
        hour12: false,
    };
    const date = new Date().toLocaleString("en-US", options);
    const result = chalk_1.grey(date);
    return result;
}
exports.generateDate = generateDate;
function createLogger(icon, color) {
    return function logger(...logs) {
        log(logs, icon, color);
    };
}
exports.createLogger = createLogger;
function createMockLogger(icon, color) {
    return function logger(...logs) {
        const date = generateDate();
        const coloredIcon = color(icon);
        return `${date} ${coloredIcon} ${logs.join(", ")}`;
    };
}
exports.createMockLogger = createMockLogger;
